File: src/main/scala/s/template_method/like/functional/DisplayFactory.scala
package s.template_method.like.functional

object DisplayFactory {

  def createString(string: String): Display = {
    val width = string.length

    def printLine(): Unit = {
      Console.print("+")
      for (_ <- 0 until width) {
        Console.print("-")
      }
      Console.println("+")
    }

    val opener = () => printLine()
    val printer = () => Console.println("|" + string + "|")
    val closer = () => printLine()

    new Display(opener, printer, closer)
  }

  def createChar(ch: Char): Display = {
    val opener = () => Console.print("<<")
    val printer = () => Console.print(ch)
    val closer = () => Console.println(">>")
    new Display(opener, printer, closer)
  }

}

File: src/main/java/j/template_method/like/functional/DisplayFactory.java
package j.template_method.like.functional;

public final class DisplayFactory {
  public static Display createString(String str) {
    return new Display(
        () -> {
          System.out.print("<<");
          return null;
        },
        () -> {
          System.out.print(str);
          return null;
        },
        () -> {
          System.out.println(">>");
          return null;
        });
  }

  public static Display createChar(char ch) {
    return new Display(
        () -> {
          System.out.print("<<");
          return null;
        },
        () -> {
          System.out.print(ch);
          return null;
        },
        () -> {
          System.out.println(">>");
          return null;
        });
  }
}
